// react-native-chart-kit/dist/v2/react-native/charts/line/seriesStyles.js
import { getLineChartAreaFillConfig, getLineChartDotConfig, getLineChartStrokeStyle, getLineChartThresholdStyle } from "./options";
import { getSeriesColor } from "./utils";
export const buildLineChartSeriesStyleMap = ({ areaFill, dots, resolvedTheme, seriesInput, showDots }) => new Map(seriesInput.map((item, index) => {
    const color = item.color ?? getSeriesColor(resolvedTheme, index);
    return [
        String(item.key ?? item.yKey),
        {
            strokeWidth: item.strokeWidth ?? 3,
            strokeStyle: getLineChartStrokeStyle({
                strokeDasharray: item.strokeDasharray,
                strokeLinecap: item.strokeLinecap,
                strokeLinejoin: item.strokeLinejoin,
                strokeOpacity: item.strokeOpacity
            }),
            area: item.area,
            areaFill: getLineChartAreaFillConfig({
                areaFill: item.areaFill ?? areaFill,
                seriesColor: color
            }),
            curve: item.curve,
            color,
            threshold: getLineChartThresholdStyle({
                seriesColor: color,
                threshold: item.threshold
            }),
            dot: getLineChartDotConfig({
                dots,
                seriesDot: item.dot,
                showDots
            })
        }
    ];
}));
